Top command

nmalocic opened the pull request:
I have added extra method to create JSON endpoint for contributors in:
GitHubController > ContributorsInformationApi that we will call from Top command to get data.

Created Frtiz.Models to hold GitHubContributor and GitHubInformation models that are used in both Frtiz.StreamTools nad Fritz.ChatBot (not sure if this is correct approach)

Created top command that hits ContributorsInformationApi  with HttpClient ( is there better way, if our HostedService is running in the context of webapplication?) and gets GitHubInformation and creates formatted message and returns it to chat.

ATM there are no tests, there should be at least some tests, but i'm not sure if i am going in right direction

Commit:
Created model project for shared models betwen StreamTools and ChatBot. Exposed JSON endpoint for contributor information, added Top command

Commit:
Removed extra config

PatrykOlejniczak commented:
Hello @nmalocic,
I noticed that some files are missing, you probably forgot to track and add them (entire Fritz.Models project files .csproject and classes GitHubInformation etc). Please add them ;)
Regards

Commit:
Fixed messup with Fritz.Models project (i hope)

nmalocic commented:
Hi @PatrykOlejniczak  thx for noticing this, for some reason new project that i added i didn't added to Fritz.StreamTools folder, so it wasn't versioned by git, i delete that project, and created new one, so i hope now it's ok.

csharpfritz commented:
I like that you are moving the GitHub models into a separate project, but I think we can re-use the Fritz.StreamLib.Core project for this.  Just create a Models folder in that project and copy the GitHub models into there.  Then we don't need to Models project.

Also, let's wrap the HttpClient creation in TopCommand.Execute in a using statement so that it is properly disposed.  

Otherwise, looks like a good command to add to the toolset.

parithon commented:
Putting the HttpClient within a using statement would result in the following issue: https://aspnetmonsters.com/2016/08/2016-08-27-httpclientwrong/

Perhaps we should create a static HttpClient used throughout the Bot.

nmalocic commented:
HI @parithon great read i must admit that i didn't knew about this issue. 
If @csharpfritz agrees i'll try to create public static HttpClient, on FritzBot.cs (or somewhere else) and we can reuse it in commands.

Commit:
resolved merge conflict changes

Commit:
Removed Models project, models are moved to Fritz.StreamLib.Core project to be reused accros projects

Commit:
Added static http client in FritzBot to be shared in other commands

Commit:
First take of using HttpClientFactory for Topcommand

nmalocic commented:
Hi @csharpfritz  i have changed TopCommand and tried to make use of DI and httpclientfactory.
As for the Image command, it is also created manually from the code so i didn't dare change that :(

I tried to copy your way of NamedClients as for Proxy class, i had some issues pulling IHttpClientFactory so decided not to do it that way

parithon commented on Fritz.StreamTools.sln:
```diff
@@ -3,10 +3,6 @@ Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 15
 VisualStudioVersion = 15.0.27130.2020
 MinimumVisualStudioVersion = 10.0.40219.1
-Project("{E53339B2-1760-4266-BCC7-CA923CBCF16C}") = "docker-compose", "docker-compose.dcproj", "{520B1405-AC30-493A-8F04-6AA9B8FE2CDD}"
```
Careful. It looks like we've removed the docker-compose and test projects.

nmalocic replied:
Hmm i also noticed that i don't have test project in same solution... :(
I think i broke it i'll investigate

nmalocic committed:
Rverted changes to .sln file where i removed projects/composer file

csharpfritz commented:
Is this ready to merge?  What's the status?

nmalocic commented:
Hi @csharpfritz i have changed HttpClient to be used from DI and moved models to different project.
My biggest problem is i don't know how to test if this works.
Other than that i think it's done.

PS.
Sorry for not beeing up to date lately

Commit:
Merge branch 'dev' of https://github.com/csharpfritz/Fritz.StreamTools into dev

Commit:
Fixed merge issues, changes to top command to reflect new interfaces

nmalocic commented:
Hey @csharpfritz,
i have resolved conflicts, and changed Top command to match the new Interfaces.

cc @parithon

csharpfritz commented:
I see a few issues that need to be addressed before we merge.  Look for my comments inline

csharpfritz commented on Fritz.Chatbot/Commands/TopCommand.cs:
```diff
@@ -0,0 +1,68 @@
+				AppendContributors(builder, repoInfo.TopEverContributors);
+			}
+
+			return message;
```
The message content is never set here to the builder.ToString()

csharpfritz commented on Fritz.StreamTools/Views/GitHub/Contributor_h-scroll.cshtml:
```diff
@@ -39,59 +40,66 @@
 		</span>
 
 	}
+>>>>>>> 51f144d29a8b42000ebd8d2a641f0517a7aaa6f1
```
Why are these merge markers included in the PR?

csharpfritz commented:
Can we also get a unit test to go along with this?

Finally, I'd like the top command to require a project name so that we do not flood chat

Commit:
Fixed merge issue, removed resolve commit notes from code

Commit:
Assigned value of the meessage

nmalocic commented:
Hey @csharpfritz ,

sorry for the slow response, i "moved" to USA for 3 month work assignment so was little jet lagged.

I have fixed comments for merge issues, and assigned value to message.
When you say "top command should require a project name", should i make that as a constructor param or on every method call?
At this point i'm not sure how to create unit test for TopCommand, i'll take a look at open PR for Unit tests and try to learn how to do it.

csharpfritz commented:
Let's make the project name and also the time range a constructor argument to the command.

We should be able to execute a command like this:   

!top CoreWiki ThisWeek

And receive information about the top contributors this week to CoreWiki

csharpfritz commented:
Closing and declining this feature.
